### PR TITLE
2132 ge ba editing action blocks

### LINKED ISSUE #2132: Support for editing the behavior annex action language
Support for editing the behavior annex action language. 

Add support for editing the behavior action block for transitions.

Editing action language blocks will be text based. An embedded Xtext editor will be integrated into the Eclipse Property view when using the OSATE graphical editor. This will allow editing without needing to transition into the text editor. This behavior is similar to the capabilities provided by other tools.
The Xtext editor will not be integrated into the diagram view of the graphical editor at this time. Depending on resource availability and technical feasibility, this can be reevaluated later on in the project.

### PR DESCRIPTION
**Pull request by RyanMcilnay**
Closes #2132

### CODE REVIEW

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java [hunk 1]
```diff
@@ -95,56 +101,199 @@ public void createControls(final Composite parent, final TabbedPropertySheetPage
...
+
+		final GridData gd = GridDataFactory.swtDefaults().align(SWT.FILL, SWT.FILL).grab(true, true)
+				.hint(SWT.DEFAULT, SWT.DEFAULT).create();
+		conditionEditingControls = new EmbeddedTextEditor(container, SWT.NONE, SWT.BORDER | SWT.SINGLE,
+				gd);
+		// Set layout data for the composite
+		conditionEditingControls.setLayoutData(gd);
+
+		final Label actionLabel = new Label(container, SWT.NONE);
+		actionLabel.setText("Action: ");
```
Comment (philip-alldredge) [inline ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java:115]: Remove unnecessary whitespace in label

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java [hunk 2]
```diff
@@ -95,56 +101,199 @@ public void createControls(final Composite parent, final TabbedPropertySheetPage
 		container.setLayout(GridLayoutFactory.swtDefaults().numColumns(2).create());
 
 		final Label conditionLabel = new Label(container, SWT.NONE);
-		conditionLabel.setText("Condition:");
+		conditionLabel.setText("Condition: ");
```
Comment (philip-alldredge) [inline ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java:104]: Remove unnecessary whitespace in label

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EditEmbeddedTextDialog.java [hunk 3]
```diff
@@ -51,34 +53,51 @@
...
 	private final IHandlerService service;
+	private final int styledTextStyle;
+	private final GridData styledTextLayoutData;
 	private IHandlerActivation undoHandler;
 	private IHandlerActivation redoHandler;
 	private StyledText styledText;
-	private String replacementText;
-
+	private String modifiedSrc;
+	private Pair<String, String> modifiedSrcToEnteredValue;
```
Comment (philip-alldredge) [inline ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EditEmbeddedTextDialog.java:73]: Remove. Doesn't seem to serve a purpose.

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EditEmbeddedTextDialog.java [hunk 4]
```diff
@@ -134,6 +153,7 @@ public void create() {
 	// Text modification listener that sets the OK button as enabled
 	// or disabled based on if the new text is valid
 	private ExtendedModifyListener getTextValidator(final BehaviorTransition behaviorTransition,
```
Comment (philip-alldredge) [inline ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EditEmbeddedTextDialog.java:155]: Consider renaming to createTextValidator(). It creates something rather than just getting it.

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EditEmbeddedTextDialog.java [hunk 5]
```diff
@@ -204,12 +225,10 @@ public boolean close() {
 		return super.close();
 	}
 
-	public String getText() {
-		return replacementText;
+	public String getText(final boolean useModifiedSource) {
```
Comment (philip-alldredge) [inline ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EditEmbeddedTextDialog.java:228]: Remove this parameter. Doesn't appear to be needed. Everything calls this function with true.

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EmbeddedTextEditor.java [hunk 6]
```diff
@@ -24,41 +24,48 @@
...
+	private StyledText styledText;
+	private Button btn;
 	private EmbeddedXtextAdapter xtextAdapter;
+	private final int styledTextStyle;
+	private final GridData styledTextLayoutData;
 
-	public EmbeddedTextControls(final Composite parent, final int style, final int styledTextStyle, final int btnStyle,
-			final String editBtnText) {
+	public EmbeddedTextEditor(final Composite parent, final int style, final int styledTextStyle,
+			final GridData styledTextLayoutData) {
```
Comment (philip-alldredge) [inline ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EmbeddedTextEditor.java:46]: Remove this styledTextLayoutData parameter. Internal detail. Using the same value in both cases. Just grab and fill contents of container.

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java [hunk 7]
```diff
@@ -95,56 +101,199 @@ public void createControls(final Composite parent, final TabbedPropertySheetPage
...
+			final XtextResource xtextResource) {
+		final EmbeddedTextValue actionTextValue = getActionBlockTextValue(behaviorTransition, sourceText);
+		final SelectionAdapter editActionBlockSelectionAdapter = getEditActionSelectionAdapter(project, actionTextValue,
+				behaviorTransition, editingDomain, xtextDocument, xtextResource, sourceText);
+		createEditingControls(actionBlockEditingControls, editActionBlockSelectionAdapter, project, actionTextValue);
+	}
+
+	private void refreshEditControls() {
+		// Dispose before refresh
+		disposeEditControls();
```
Comment (philip-alldredge) [inline ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java:189]: Instead of having to dispose things manually before refreshing, the refresh() of the editing controls should do this.

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EmbeddedTextEditor.java [hunk 8]
```diff
@@ -24,41 +24,48 @@
...
 		super(parent, style);
 		this.setBackground(parent.getBackground());
 		this.setLayout(GridLayoutFactory.swtDefaults().numColumns(2).create());
-		this.setLayoutData(GridDataFactory.swtDefaults().align(SWT.FILL, SWT.FILL).grab(true, true)
-				.hint(SWT.DEFAULT, SWT.DEFAULT).create());
+		this.styledTextStyle = styledTextStyle;
+		this.styledTextLayoutData = styledTextLayoutData;
+	}
 
+	public void refresh() {
```
Comment (philip-alldredge) [inline ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EmbeddedTextEditor.java:54]: Consider renaming to createControls(). Document when this should be called. Consider allowing specifying the testing IDs as parameters. It is odd that must call set*TestId multiple times. This class is leaking what should be internal details into its API. Alternatively, the class could store the testing IDs and automatic apply and reapply them as needed. (Shouldn't matter when set*TestId is called but it should only need to be called once).

**Review by philip-alldredge — CHANGES_REQUESTED** [review]
In addition to the comments made to the code., the action isn't undoable when I modify a dispatch condition or the action block without the text editor being open.

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java [hunk 9]
```diff
@@ -95,56 +101,199 @@ public void createControls(final Composite parent, final TabbedPropertySheetPage
...
+							// Get source text for xtext document
+							final String srcText = dlg.getText(true);
+							// Execute modification with xtext document
+							actionService.execute("Modifying Behavior Transition Action Block", ExecutionMode.NORMAL,
+									new EmbeddedTextModificationAction(xtextDocument, modelChangeNotifier, project,
+											srcText));
+						} else {
+							// Execute modification with xtext resource
+							final boolean actionExists = !actionTextValue.getEditableText().isEmpty();
+							final String newText = getNewText(dlg.getText(false), actionExists);
```
Comment (RyanMcilnay) [inline ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java:252]: getText(boolean) is called with false here for editing the XtextResource (editor closed).

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java [hunk 10]
```diff
@@ -159,25 +308,35 @@ public void widgetSelected(final SelectionEvent e) {
...
+						if (xtextDocument != null) {
+							// Execute modification with xtext document
+							actionService.execute("Modifying Behavior Transition Condition", ExecutionMode.NORMAL,
+									new EmbeddedTextModificationAction(xtextDocument, modelChangeNotifier, project,
+											dlg.getText(true)));
+						} else {
+							// Execute modification with xtext resource
+							actionService.execute("Modifying Behavior Transition Condition", ExecutionMode.NORMAL,
+									new EmbeddedTextModificationAction(editingDomain, xtextResource,
+											modelChangeNotifier, project, dlg.getText(false),
```
Comment (RyanMcilnay) [inline ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java:330]: getText(boolean) is also called with false here for editing the XtextResource (editor closed).

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EditEmbeddedTextDialog.java (hunk 5 above)
Comment (RyanMcilnay) [reply ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EditEmbeddedTextDialog.java:228]: I call true when editing the XtextDocument, to replace entire text using doc.set(newSrc).  And I use false when I am just replacing a part of the text in the XtextResource.

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EditEmbeddedTextDialog.java (hunk 3 above)
Comment (RyanMcilnay) [reply ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EditEmbeddedTextDialog.java:73]: Refer to other comments about why this exists.  We can talk about better ways to do this though.  I return the value from the styled text when editing XtextResources and the entire source for modifying XtextDocuments.

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EmbeddedTextEditor.java (hunk 8 above)
Comment (RyanMcilnay) [reply ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/EmbeddedTextEditor.java:54]: Good point.  I fixed this.

File: ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java (hunk 9 above)
Comment (philip-alldredge) [reply ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java:244]: Okay, I missed that . if that is need then that's fine. In general I was confused a bit about all the modified text stuff. It wasn't clear what some of that was doing.
